internal/controller/serving/reconcilers/kserve_keda_reconciler.go

```diff
@@ -0,0 +1,629 @@
+	"github.com/kuadrant/authorino/pkg/log"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	kedaapi "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+)
+
+const (
+	KEDAResourcesPrefix                            = "inference-"
+	KEDAPrometheusPrefix                           = KEDAResourcesPrefix + "prometheus-"
+	KEDAPrometheusAuthResourceName                 = KEDAPrometheusPrefix + "auth"
+	KEDAPrometheusAuthServiceAccountName           = KEDAPrometheusAuthResourceName
+	KEDAPrometheusAuthTriggerSecretName            = KEDAPrometheusAuthResourceName
+	KEDAPrometheusAuthMetricsReaderRoleName        = KEDAPrometheusAuthResourceName
+	KEDAPrometheusAuthMetricsReaderRoleBindingName = KEDAPrometheusAuthResourceName
+	KEDAPrometheusAuthTriggerAuthName              = KEDAPrometheusAuthResourceName
+
+	KEDAResourcesLabelKey   = "odh-model-controller"
+	KEDAResourcesLabelValue = "keda-reconciler"
+)
+
+var (
+	KedaLabelPredicate predicate.Predicate
+)
+
+func init() {
+	p, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			KEDAResourcesLabelKey: KEDAResourcesLabelValue,
+		},
+	})
+	if err != nil {
+		panic("failed to create label selector predicate: " + err.Error())
+	}
+	KedaLabelPredicate = p
+}
+
+var _ SubResourceReconciler = (*KserveKEDAReconciler)(nil)
+
+// KserveKEDAReconciler allows ISVCs to autoscale on custom Prometheus metrics via KEDA, with secure OpenShift Monitoring
+// access. The reconciler automates KEDA/RBAC resource lifecycle.
+//
+// KServeKEDAReconciler manages KEDA-specific resources (ServiceAccount, Secret, Role, RoleBinding, TriggerAuthentication)
+// for Prometheus-based autoscaling.
+// - Creates resources if InferenceService uses KEDA Prometheus external metric.
+// - Adds each InferenceService in a given namespace as non-controlling owner to shared namespaced resources.
+// - Removes InferenceService owner reference if KEDA Prometheus autoscaling is unused or InferenceService deleted.
+// - Cleans up KEDA resources from namespace if no InferenceServices use KEDA Prometheus autoscaling.
+type KserveKEDAReconciler struct {
+	client client.Client
+}
+
+func NewKServeKEDAReconciler(client client.Client) *KserveKEDAReconciler {
+	return &KserveKEDAReconciler{
+		client: client,
+	}
+}
+
+func (k *KserveKEDAReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+
+	log = log.WithName("KserveKEDAReconciler")
+	log.V(2).Info("Reconciling InferenceService", "InferenceService", isvc)
+
+	if !hasPrometheusExternalAutoscalingMetric(isvc) {
```

Comment:
so, it is only enabled when the autoscaler has Prometheus-based metrics?

Reply:
Yes, in the other cases, we don't really need it